runtime/compiler/trj9/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,205 @@
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#include "JProfilingRecompLoopTest.hpp"
+#include "il/Block.hpp"
+#include "infra/Cfg.hpp"
+#include "infra/TRCfgEdge.hpp"
+#include "infra/Checklist.hpp"
+#include "infra/ILWalk.hpp"
+#include "infra/List.hpp"
+#include "optimizer/Optimization_inlines.hpp"
+#include "il/Node_inlines.hpp"
+#include "infra/Checklist.hpp"             // for TR::NodeChecklist
+#include "ras/DebugCounter.hpp"
+#include "control/Recompilation.hpp"              // for TR_Recompilation, etc
+#include "control/RecompilationInfo.hpp"              // for TR_Recompilation, etc
+#include "codegen/CodeGenerator.hpp"
+#include "optimizer/TransformUtil.hpp"            // for TransformUtil
+#include "optimizer/JProfilingBlock.hpp"
+
+
+/**
+ * Walks the TR_RegionStructure counting loops to get the nesting depth of the block
+ */ 
+int32_t TR_JProfilingRecompLoopTest::getLoopNestingDepth(TR::Compilation *comp, TR::Block *block)
+   {
+   TR_RegionStructure *region = block->getParentStructureIfExists(comp->getFlowGraph());
+   int32_t nestingDepth = 0;
+   while (region && !region->isAcyclic())
+      {
+      nestingDepth++;
+      region = region->getParent();
+      }
+   return nestingDepth;
+   }
+
+
+int32_t
+TR_JProfilingRecompLoopTest::perform()
+   {
+   if (comp()->getOption(TR_EnableJProfiling))
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled, run JProfiling\n");
+      }
+   else if (comp()->getProfilingMode() == JProfiling)
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled for profiling compilations, run JProfilingRecompLoopTest\n");
+      }
+   else
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has not been enabled, skip JProfilingRecompLoopTest\n");
+      return 0;
+      }
+   RecompilationTestLocations testLocations(RecompilationTestLocationAllocator(comp()->trMemory()->currentStackRegion()));
+   TR::TreeTop *cursor = comp()->getStartTree();
+   TR::CFG *cfg = comp()->getFlowGraph();
+   TR::Block *currentBlock = NULL;
+   while (cursor)
+      {
+      if (cursor->getNode()->getOpCodeValue() == TR::BBStart)
+         {
+         currentBlock = cursor->getNode()->getBlock();
+         }
+      else if (cursor->getNode()->getOpCodeValue() == TR::asynccheck)
+         {
+         TR_ASSERT_FATAL(currentBlock!=NULL,"We should have encountered BBStart before and should have the enclosing block");
+         if (currentBlock->getStructureOf()->getContainingLoop() != NULL)
```

Comment:
@andrewcraik  Regarding to this comment https://github.com/eclipse/openj9/pull/1601#discussion_r178960155
I was checking the function which is called, [getContainingLoop()](https://github.com/eclipse/omr/blob/1c979dbe7b3251bf0cad1f5dd375083ae862cefb/compiler/optimizer/Structure.cpp#L106), Looks like it walks the parent region structure till encountering `NULL` to get the natural loop if found otherwise returns NULL which is what is checked here.